### ISSUE TITLE
Parse error: syntax error, unexpected '' (T_ENCAPSED_AND_WHITESPACE), expecting end of file in ../vendor/composer/autoload_namespaces.php on line 14

### ISSUE THREAD

**Issue by YspMark** [composer/composer#8608]
Composer gets broken at a `composer install` after switching branches in my project. It is giving a `Parse error: syntax error, unexpected '' (T_ENCAPSED_AND_WHITESPACE), expecting end of file in /projects/dashboard/vendor/composer/autoload_namespaces.php on line 14` at the Generating autoload files step of the install.

Since it is indicated at the autoload file I tried to `composer dump-autoload` resulting in the exact same error.

I can only fix this by removing the entire /vendor/composer folder and `composer install` after.
The resulting autoload_namespaces.php file is however exactly the same after this.

Output of `composer diagnose`:

```
Checking composer.json: WARNING
No license specified, it is recommended to do so. For closed-source software you may use "proprietary" as license.
Checking platform settings: OK
Checking git settings: OK
Checking http connectivity to packagist: OK
Checking https connectivity to packagist: OK
Checking github.com rate limit: OK
Checking disk free space: OK
Checking pubkeys: 
Tags Public Key Fingerprint: 57815BA2 7E54DC31 7ECC7CC5 573090D0  87719BA6 8F3BB723 4E5D42D0 84A14642
Dev Public Key Fingerprint: 4AC45767 E5EC2265 2F0C1167 CBBB8A2B  0C708369 153E328C AD90147D AFE50952
OK
Checking composer version: OK
Composer version: 1.9.3
PHP version: 7.3 - Package overridden via config.platform (actual: 7.3.14)
PHP binary path: /usr/bin/php7.3
```

When I run this command:

```
composer install

composer dump-autoload
```

I get the following output:

```
...
Generating autoload files

Parse error: syntax error, unexpected '' (T_ENCAPSED_AND_WHITESPACE), expecting end of file in /projects/dashboard/vendor/composer/autoload_namespaces.php on line 14

```

Contents of `autoload_namespaces.php`:

```
<?php

// autoload_namespaces.php @generated by Composer

$vendorDir = dirname(dirname(__FILE__));
$baseDir = dirname($vendorDir);

return array(
    'Rize\\UriTemplate' => array($vendorDir . '/rize/uri-template/src'),
    'Pimple' => array($vendorDir . '/pimple/pimple/src'),
    'Google_Service_' => array($vendorDir . '/google/apiclient-services/src'),
    'Google_' => array($vendorDir . '/google/apiclient/src'),
);
```

**Comment by Seldaek** [comment]
Is the file you posted the one *after* you delete all vendor and run install again? Could you upload the file from before, the one that breaks for sure, somewhere where it can be uploaded as a file and not copy the content out, because it seems like there is something weird at the end of the file which is not visible here..

**Comment by YspMark** [comment]
Both before and after are the same, verified with a md5sum over both files:
https://gist.github.com/YspMark/8d52e1ef27e9a2a6a0043ee6d7abba25

**Comment by Seldaek** [comment]
But it says `on line 14`, that file has 13 lines only, and it passes a `php -l` test so it's valid.. There must be something else at play here.

**Comment by YspMark** [comment]
That is exactly why I posted, this is the literal error I got from `composer install` and couldn't resolve the issue other than delete the /vendor/composer folder and reinstall with composer. So I presume it is an issue with composer (internally).

**Comment by Seldaek** [comment]
But that error comes from PHP though.. Can you run `dump-autoload` again (or whichever command fails) with `--no-plugins -vvv` added to it, see if it still fails? And if it does please paste full output and if possible share the composer.json or a subset of it that still causes the issue.. Because I have no idea what it can be right now.

**Comment by YspMark** [comment]
Output from `composer dump-autoload --no-plugins -vvv`

```
Running 1.9.3 (2020-02-04 12:58:49) with PHP 7.3.14-5+ubuntu16.04.1+deb.sury.org+2 on Linux / 4.4.0-96-generic
Reading ./composer.json
Loading config file ./composer.json
Checked CA file /etc/ssl/certs/ca-certificates.crt: valid
Executing command (/projects/dashboard): git branch --no-color --no-abbrev -v
Reading /data/www/.composer/composer.json
Loading config file /data/www/.composer/composer.json
Reading /projects/dashboard/vendor/composer/installed.json
Generating autoload files
Parse error: syntax error, unexpected '' (T_ENCAPSED_AND_WHITESPACE), expecting end of file in /projects/dashboard/vendor/composer/autoload_namespaces.php on line 14
```

composer.json:
```
{
  "require": {
    "ext-curl": "*",
    "ext-json": "*",
    "guzzlehttp/guzzle": "~6.0",
    "slim/twig-view": "^2.4",
    "yoursurprisecom/apm": "^2.0",
    "yoursurprisecom/dashboard-api": "^5.0"
  },
  "require-dev": {
    "yoursurprisecom/deployer-recipes": "^1.0"
  },
  "repositories": [
    {
      "type": "vcs",
      "url": "git@gitlab.<<private_repo_url>>/dashboard-api.git"
    },
    {
      "type": "vcs",
      "url": "git@gitlab.<<private_repo_url>>/deployer-recipes.git"
    },
    {
      "type": "vcs",
      "url": "git@gitlab.<<private_repo_url>>/apm.git"
    }
  ],
  "config": {
    "sort-packages": true,
    "platform": {
      "php": "7.3"
    }
  },

  "minimum-stability": "stable",
  "prefer-stable": true
}
```
Also an `composer install --no-plugins -vvv` doesn't fix the issue and generates the same error at the Generating autoload files step.

**Comment by Seldaek** [comment]
Ok, well I can't reproduce this here, I am not sure if it's because I can't install your private deps or because the issue comes from something else though..

I imagine that the point triggering the failing file include is https://github.com/composer/composer/blob/fce60aadd17230ba42fb3e76a8588f3d170da527/src/Composer/Autoload/AutoloadGenerator.php#L767 - in case you want to run composer from source and try to figure this out. Maybe doing a `var_dump(file_get_contents($targetDir . '/autoload_namespaces.php'));` at that point would yield some clue.. 

Maybe also try to run with opcache disabled, who knows that it'd be a weird bug there.

**Comment by YspMark** [comment]
Didn't manage to run composer from source to try your suggestion. However today I installed and removed another package and got served a different error leading me to a possible cause.
The error I got:
```
Writing lock file
Generating autoload files

Parse error: syntax error, unexpected '/src', ' in /projects/dashboard/vendor/composer/autoload_psr4.php on line 68
```
vendor/composer/autoload_psr4.php:
https://gist.github.com/f50189929ac8b5d714263bbc90e4b5fe

Looks like slim-session at line 68 isn't using a namespace but also creates an unclosed array in the autoload files. And I assume triggering the parse errors.

**Comment by Seldaek** [comment]
I'm not sure what you mean, the array is closed and syntax looks all fine (it doesn't error in https://3v4l.org/PH5bA). Something seems deeply off with your php or filesystem or something..

As for the empty namespace this is a supported feature for fallback dirs where every class will be looked up.

**Comment by YspMark** [comment]
I think I misread the array.

I'am starting to beleave my filesystem or php in general is off, since opcache is not loaded. Weird that only Composer is having issues, so might only be the php-cli that is bugged.

**Comment by alcohol** [comment]
Can you share more about your environment? Are you using containers? Virtual machines? Network shares/directories?

**Comment by YspMark** [comment]
I'am running inside a Vagrant box, based on Ubuntu 16.04 running on Linux host-machine in Virtualbox 6.0.16. The source-code is inside a synced_folder.
I run composer from within that box, switching branches with Git inside or outside the box doesn't matter both give same result regarding the issue. So that tends to rule out a possible syncing issue with vagrant?

**Comment by YspMark** [comment]
Had some colleagues testing the same procedure, most don't face the issue. Only another colleague at MacOS using the same Virtualbox version, so begins to look like a Virtualbox 6.0.16 issue.

Thanks for the patience and help.

**Comment by alcohol** [comment]
VirtualBox is known to cause filesystem issues or issues that relate to the filesystem in general. Avoid using synced directories between host and VM. But even just inside the VM it can be a bit of a hit and miss scenario.

**Comment by YspMark** [comment]
Downgraded to VirtualBox 5.2.36 and issue is gone, very confusing since only composer seems to be affected by this.

**Comment by Black0wI** [comment]
Hi, 

I managed to reproduce and fix this. 

Config : Vagrant 2.2.5 , VBox 6.0.20, PHP 7.3.17-1+0~20200419.57+debian9~1.gbp0fda17 (cli), sur Mac OS Catalina.

my original `composer/autoload_namespaces.php` file was :


```php
// autoload_namespaces.php @generated by Composer

$vendorDir = dirname(dirname(__FILE__));
$baseDir = dirname($vendorDir);

return array(
    'ZfBugsnag' => array($vendorDir . '/nickurt/zf-bugsnag/src'),
    'ProxyManager\\' => array($vendorDir . '/ocramius/proxy-manager/src'),
    'Mailgun' => array($vendorDir . '/mailgun/mailgun-php/src'),
    'JsonMapper' => array($vendorDir . '/netresearch/jsonmapper/src'),
    'Google_Service_' => array($vendorDir . '/google/apiclient-services/src'),
    'Google_' => array($vendorDir . '/google/apiclient/src'),
    'DoctrineORMModule\\' => array($vendorDir . '/doctrine/doctrine-orm-module/src'),
    'DoctrineModule\\' => array($vendorDir . '/doctrine/doctrine-module/src'),
    'Bugsnag_' => array($vendorDir . '/bugsnag/bugsnag/src'),
);
```
 And throws the same error : 
```shell
Parse error: syntax error, unexpected '' => array($vendorDir . '/bugs' (T_ENCAPSED_AND_WHITESPACE) in /var/www/html/vendor/composer/autoload_namespaces.php on line 19
```

I fixed this by removing the last coma at the end of the last value of array : 
```php
    'Bugsnag_' => array($vendorDir . '/bugsnag/bugsnag/src')
);
```

No other explanation...

**Comment by mschmitz89** [comment]
Thank you @Black0wI 
This fixed the problem for me, too!

**Comment by aleksandrovskaya** [comment]
I'm also using VirtualBox 6.0.24 (vagrant) + Ubuntu guest `5.4.0-77-generic #86-Ubuntu` environment with synced folder and meet the same problem.
I've noticed that doing:
```
$ sudo su
# sync; echo 1 > /proc/sys/vm/drop_caches
```
helps. This drops all page cache + doing writeback on the disk.

Seems like we have a bug in VirtualBox vboxsf driver.

**Comment by Michael-Wohlstadter** [comment]
Mostly the same behavior.  I had been deleting the comma per BlackOwl but this morning was prompted for an equals sign.  I opened the autoload file and there was no equals sign and indeed the line reference was one line past the end of the file.

So I just typed a space then backspaced over it (i.e. no actual change to the file).  Then I saved it and ran "composer update" without error.

My environment:

- Composer 2.1.14
- PHP 7.3.33
- Vagrant 2.2.19
- VirtualBox 6.1.30 r148432
- macOS 10.15.7 host
- Ubuntu 20.04.3 (Linux 5.4.0-91-generic) guest

Vagrant / VirtualBox combination is a common thread.
